### PR TITLE
Make all paths relative in `kedro catalog resolve`

### PR DESCRIPTION
**Pull request by ankatiyar**
## Description

This bug was my bad! 😅 
I'd suggested in #3561 to just remove the trimming of filepaths because we were only checking for `filepath` key in the catalog config but `PartitionedDataset` uses the key `path` which was causing an error.

But the filepaths in `context.catalog._dataset_patterns` are absolute paths (because of this https://github.com/kedro-org/kedro/blob/e8f1bfd72992336ec12591b49a5fa2654217472f/kedro/framework/context/context.py#L54) . So when running `kedro catalog resolve`, the  output for factory datasets had absolute paths but explicit datasets were all relative paths 

Eg - 
```yaml
"{pattern}":
  type: partitions.PartitionedDataset
  path: data/01_raw
  dataset: pandas.CSVDataset
  metadata:
    path:  data/01_raw/metadata.yml # just to see what happens with nested dicts
```
* `kedro catalog resolve`
```yaml
y_train: # resolved from factory
  dataset: pandas.CSVDataset
  metadata:
    path: /Users/ankita_katiyar/kedro_projects/space/data/01_raw/metadata.yml
  path: /Users/ankita_katiyar/kedro_projects/space/data/01_raw
  type: partitions.PartitionedDataset

reviews: # explicit dataset entry
  filepath: data/01_raw/reviews.csv
  type: pandas.CSVDataset
```

## Development notes

- trimming of filepath now works for nested stuff also 
- refactored the tests a little bit, still haven't fully gotten rid of the mocking of the Context but it's testing the trimming .

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

File: kedro/framework/cli/catalog.py [hunk 1]
```diff
@@ -264,7 +264,21 @@ def resolve_patterns(metadata: ProjectMetadata, env: str) -> None:
...
+            explicit_datasets[ds_name] = ds_config_trimmed
 
     secho(yaml.dump(explicit_datasets))
+
+
+def _trim_filepath(project_path: str, config: dict) -> dict:
+    """Trim the filepaths in the config dictionary, make them relative to the project path."""
+    conf_keys_with_path = ["filepath", "path", "filename"]
+    for key, value in config.items():
+        if isinstance(value, dict):
```
Comment (AhdraMeraliQB) [inline kedro/framework/cli/catalog.py:280]: Nice recursive case!

**Review by AhdraMeraliQB — APPROVED** [review]
Awesome work with the tests! 🌟

**Review by noklam — COMMENTED** [review]
Actually I have the question, do we want to show absolute path or relative path? Resolving to an absolute path doesn't seem wrong to me. Is this consistent to `kedro catalog list`?

From my understanding, we are basically trying to undo this line of code.
https://github.com/kedro-org/kedro/blob/941440aa4a5815e8d8157d167ec581524a1f6e20/kedro/framework/context/context.py#L99

Would it be easier to actually create the DataCatalog without going `context.catalog` instead of creating a recursive call to undo this?

**Comment by ankatiyar** [comment]
@noklam you're right, I've updated it!

**Review by merelcht — APPROVED** [review]
Tested and it works well! 👍

**Pull request merged**